app/src/server/tasks/getTestResult.task.ts

```diff
@@ -0,0 +1,174 @@
+import { type JsonValue } from "type-fest";
+import { type ChatCompletionMessage } from "openai/resources/chat";
+import { APIError } from "openai/error";
+
+import { prisma } from "~/server/db";
+import hashObject from "~/utils/hashObject";
+import defineTask from "./defineTask";
+import {
+  pruneInputMessages,
+  getCompletion,
+  getStringsToPrune,
+} from "~/modelProviders/fine-tuned/getCompletion";
+import { countLlamaChatTokens, countLlamaChatTokensInMessages } from "~/utils/countTokens";
+import { getCompletion2 } from "~/modelProviders/fine-tuned/getCompletion-2";
+
+export type GetTestResultJob = {
+  fineTuneId: string;
+  datasetEntryId: string;
+  numPreviousTries: number;
+};
+
+const MAX_AUTO_RETRIES = 50;
+const MIN_DELAY = 500; // milliseconds
+const MAX_DELAY = 15000; // milliseconds
+
+function calculateDelay(numPreviousTries: number): number {
+  const baseDelay = Math.min(MAX_DELAY, MIN_DELAY * Math.pow(2, numPreviousTries));
+  const jitter = Math.random() * baseDelay;
+  return baseDelay + jitter;
+}
+
+export const getTestResult = defineTask<GetTestResultJob>("getTestResult", async (task) => {
+  const { fineTuneId, datasetEntryId, numPreviousTries } = task;
+
+  const [fineTune, datasetEntry] = await prisma.$transaction([
+    prisma.fineTune.findUnique({
+      where: { id: fineTuneId },
+    }),
+    prisma.datasetEntry.findUnique({
+      where: { id: datasetEntryId },
+    }),
+  ]);
+
+  if (!fineTune || !datasetEntry) return;
+
+  const existingTestEntry = await prisma.fineTuneTestingEntry.findUnique({
+    where: { fineTuneId_datasetEntryId: { fineTuneId, datasetEntryId } },
+  });
+
+  if (existingTestEntry?.output) return;
+
+  let prunedInput = existingTestEntry?.prunedInput;
+
+  const stringsToPrune = await getStringsToPrune(fineTune.id);
+  if (!existingTestEntry) {
+    prunedInput = pruneInputMessages(
+      datasetEntry.input as unknown as ChatCompletionMessage[],
+      stringsToPrune,
+    );
+    await prisma.fineTuneTestingEntry.create({
+      data: {
+        fineTuneId,
+        datasetEntryId,
+        prunedInputTokens: countLlamaChatTokens(prunedInput),
+        prunedInput: prunedInput,
+      },
+    });
+  }
+
+  const cacheKey = hashObject({
+    fineTuneId,
+    input: prunedInput,
+  } as JsonValue);
+
+  const matchingTestEntry = await prisma.fineTuneTestingEntry.findFirst({
+    where: {
+      cacheKey,
+    },
+  });
+
+  if (matchingTestEntry?.output) {
+    await prisma.fineTuneTestingEntry.update({
+      where: { fineTuneId_datasetEntryId: { fineTuneId, datasetEntryId } },
+      data: {
+        output: matchingTestEntry.output,
+        outputTokens: matchingTestEntry.outputTokens,
+        errorMessage: null,
+      },
+    });
+    return;
+  }
+
+  let shouldRetry = false;
+  let completion;
+  const input = {
+    model: `openpipe:${fineTune.slug}`,
+    messages: datasetEntry.input as unknown as ChatCompletionMessage[],
+  };
+  try {
```

Comment:
Can we extract out the API code and reuse it in both places? Basically I think we should have a function that looks something like `getCompletion(fineTune: FineTune, input: { prompt: string, max_tokens: int, ... }`. That function can figure out the pipeline version, etc.

Just think it's quite likely we'll forget to this code in both places as we update the pipeline.

Reply:
I don't see a good way to do that because we need to handle models not being set up for inference differently in `getTestResult` and the `createChatCompletion` route.

In the case of `getTestResult`, if the model is not set up correctly, we need to update the testing entry's error message and return. In the case of `createChatCompletion`, we need to throw a TRPCError. In both cases, if an error is caused by anything other than the model not being set up for inference, we need to handle that error differently, meaning that we can't simply create a new `getWrappedCompletion` function that handles pipeline routing and throws an error if the model is not set up for inference, then handle that error the same way we're now handling these model inference set up errors - because if the error that `getWrappedCompletion` throws is not a model setup error, we need to handle that differently.

One option is to create a new `getWrappedCompletion` function that throws a error with the message "The model is not set up for inference" and then specifically check for that error message in whatever error is thrown by `getWrappedCompletion`, and handle it the way we currently do in these separate calling contexts.

Reply:
Aside from all that, I don't think we should invest too much time in pipeline-version specific code, because unless I'm mistaken we'll be replacing all version 0 models with version 1 models soon anyway.

Reply:
We'll delete all v0 code, but we'll always need pipeline management code because at some point (likely soon, there are things I think we can improve on the current design) we'll release a v2, v3, etc. and still need to support the old models.

I'd probably have the extracted function return either the completion or an error, and then let the caller decide how to expose the error to the user. But doesn't have to block this PR I guess.